franklin-crypto/src/circuit/rescue.rs

```diff
@@ -0,0 +1,271 @@
+            let mut found_one = false;
+            for b in BitIterator::new(alpha.into_repr()) {
+                found_one |= b;
+                if !found_one {
+                    continue;
+                }
+                tmp.push(b);
+            }
+            tmp
+        };
+        let squares={//define square, square of square, etc variables
+            let mut tmp:Vec<AllocatedNum<E>>=vec![x.clone()];
+            //alpha is constant
+            for i in 1..alpha_bits_be.len(){
+                let sqr=tmp.last().unwrap().square(cs.namespace(|| format!("sqr{}",i)))?;
+                tmp.push(sqr);
+            }
+            tmp
+        };
+        assert_eq!(squares.len(),alpha_bits_be.len());
+        let res={
+            let n=alpha_bits_be.len();
+            let mut tmp=squares[n-1].clone();
+            //alpha is constant
+            for i in 1..n{
+                if alpha_bits_be[i] {
+                    tmp=squares[n-1-i].mul(cs.namespace(|| format!("mul_due_to_bit{}",i)),&tmp)?;
+                }
+            }
+            tmp.clone()
+        };
+        Ok(res) 
+}
+fn generate_roots<CS,E>(
+    mut cs: CS,
+    x: &AllocatedNum<E>,
+    alphas: &(Option<E::Fr>,Option<E::Fr>)
+) -> Result<AllocatedNum<E>, SynthesisError>
+    where CS: ConstraintSystem<E>,E:Engine
+{
+        //Calculate the root
+        let result_var=AllocatedNum::alloc(cs.namespace(|| "found root"), ||{
+            let zero=BigUint::from_str("0").unwrap();
+            let one=BigUint::from_str("1").unwrap();
+            let p={
+                let mut p=E::Fr::char();
+                p.sub_noborrow(&E::Fr::one().into_repr());
+                let p=E::Fr::from_repr(p).unwrap();
+                fr_to_biguint(&p)+&one
+            };
+            let k_inv=fr_to_biguint(&alphas.1.unwrap());
+            let a=fr_to_biguint(&x.get_value().unwrap());
+            let res=a.modpow(&k_inv,&p);
+            Ok(biguint_to_fr(&res).unwrap())
+        })?;
+
+        {//Prove that the root is correct
+            let alpha_bits_be = {
+                let mut tmp = Vec::with_capacity(E::Fr::NUM_BITS as usize);
+                let mut found_one = false;
+                for b in BitIterator::new(alphas.0.unwrap().into_repr()) {
+                    found_one |= b;
+                    if !found_one {
+                        continue;
+                    }
+                    tmp.push(b);
+                }
+                tmp
+            };
+            let squares={
+                let mut tmp:Vec<AllocatedNum<E>>=vec![result_var.clone()];
+                for i in 1..alpha_bits_be.len(){
+                    let sqr=tmp.last().unwrap().square(cs.namespace(|| format!("root_result_sqr{}",i)))?;
+                    tmp.push(sqr);
+                }
+                tmp
+            };
+            assert_eq!(squares.len(),alpha_bits_be.len());
+            let n=alpha_bits_be.len();
```

Comment:
please add formulas as comments for each block of computation for readibility: document your intention.